Core/GDCore/Project/EffectsContainer.cpp

```diff
@@ -0,0 +1,140 @@
+/*
+ * GDevelop Core
+ * Copyright 2008-2016 Florian Rival (Florian.Rival@gmail.com). All rights
+ * reserved. This project is released under the MIT License.
+ */
+#include "EffectsContainer.h"
+
+#include "Effect.h"
+#include "GDCore/Serialization/SerializerElement.h"
+
+namespace gd {
+Effect EffectsContainer::badEffect;
+
+EffectsContainer::EffectsContainer() {}
+
+EffectsContainer::EffectsContainer(const EffectsContainer& other) {
+  Init(other);
+}
+
+EffectsContainer& EffectsContainer::operator=(const EffectsContainer& rhs) {
+  if (this != &rhs) Init(rhs);
+  return *this;
+}
+
+void EffectsContainer::Init(const EffectsContainer& other) {
+  effects.clear();
+  for (auto& it : other.effects) {
+    effects.push_back(it);
```

Comment:
Is this really triggering a copy of the effect? Or a copy of just the shared_ptr (in which case it will point to the same Effect object in memory)

Reply:
I'm pretty confused about that as well, I don't really see a copy constructor for `Effect`. Should I create an `Effect` object (and maybe write a copy constructor for Effect), then push it?

Reply:
Check what the type `auto& ` is here (VSCode can maybe autocomplete if you put your cursor on it).
I think it's a `std::shared_ptr<gd::Effect>&`. In other words, a reference to a shared_ptr. So far so good!
Then if you `push_back` it, then the vector will make a copy of what you pass (a `std::vector` store things by copying them). So it will copy the `shared_ptr<gd::Effect>`. So you'll have `other.effects` and `effects` both containing pointers pointing to the same things.

Instead, you must make a copy, i.e: make a new shared_ptr to a new object :) 
How to do so? 
1. Use * to get the object pointed by the shared_ptr.
2. Construct a new object from it
3. Make a shared ptr from it.

Something like:
```cpp
Effect* newEffect = new Effect();
*newEffect = *it; // This is the operator= that is used here.
effects.push_back(std::shared_ptr<Effect>(newEffect));
```

Or to avoid making an empty object and then assigning it (which is useless work):

```cpp
Effect* newEffect = new Effect(*it); // This is the copy constructor that is used.
effects.push_back(std::shared_ptr<Effect>(newEffect));
```

In practice, for conciseness and safety, we use make_shared:
```cpp
effects.push_back(std::make_shared<gd::Effect>(*it)); // make_shared will call the constructor, and then return a shared_ptr to the newly created object.
```